lib/compiler/src/sys_core_alias.erl

```diff
@@ -0,0 +1,317 @@
+
+-type sub() :: #sub{}.
+
+-spec module(cerl:c_module(), [compile:option()]) ->
+	{'ok',cerl:c_module(),[]}.
+
+module(#c_module{defs=Ds0}=Mod, _Opts) ->
+    Ds1 = [def(D) || D <- Ds0],
+    {ok,Mod#c_module{defs=Ds1},[]}.
+
+def({#c_var{name={F,Arity}}=Name,B0}) ->
+    try
+	put(new_var_num, 0),
+	{B1,_} = cerl_trees:mapfold(fun pre/2, fun post/2, sub_new(undefined), B0),
+	erase(new_var_num),
+	{Name,B1}
+    catch
+	Class:Error ->
+	    Stack = erlang:get_stacktrace(),
+	    io:fwrite("Function: ~w/~w\n", [F,Arity]),
+	    erlang:raise(Class, Error, Stack)
+    end.
+
+pre(#c_let{vars=Vars}=Node, Sub) ->
+    {Node,sub_fold(get_variables(Vars), Sub)};
+
+pre(#c_fun{vars=Vars}=Node, Sub) ->
+    {Node,sub_fold(get_variables(Vars), Sub)};
+
+pre(#c_clause{pats=Pats}=Node, Sub0) ->
+    VarNames = get_variables(Pats),
+    Sub1 = sub_fold(VarNames, Sub0),
+    PreKeys = get_pattern_keys(Pats),
+    {Sub2,Keys} = sub_add_keys(PreKeys, Sub1),
+
+    #sub{v=SubNames,t=Temp} = Sub2,
+    Sub3 = Sub2#sub{v=merge_variables(VarNames, SubNames),
+		    t={clause,Pats,Keys,SubNames,Temp}},
+
+    {Node#c_clause{pats=[]},Sub3};
+
+pre(Node, Sub0) ->
+    %% We cache only tuples and cons.
+    case cerl:is_data(Node) andalso not cerl:is_literal(Node) of
+	false ->
+	    {Node,Sub0};
+	true ->
+	    Kind = cerl:data_type(Node),
+	    Es = cerl:data_es(Node),
+	    case sub_cache_nodes(Kind, Es, Sub0) of
+		{Name,Sub1} ->
+		    {cerl:ann_c_var(cerl:get_ann(Node), Name),Sub1};
+		error ->
+		    {Node,Sub0}
+	    end
+    end.
+
+post(#c_let{}=Node, Sub) ->
+    {Node,sub_unfold(Sub)};
+
+post(#c_fun{}=Node, Sub) ->
+    {Node,sub_unfold(Sub)};
+
+post(#c_clause{}=Node, #sub{t={clause,Pats0,Keys,V,T}}=Sub0) ->
+    {Sub1,PostKeys} = sub_take_keys(Keys, Sub0),
+    Pats1 = put_pattern_keys(Pats0, PostKeys),
+    Sub2 = sub_unfold(Sub1#sub{v=V,t=T}),
+    {Node#c_clause{pats=Pats1},Sub2};
+
+post(Node, Sub) ->
+    {Node,Sub}.
+
+%% sub_new/1
+%% sub_add_keys/2
+%% sub_take_keys/3
+%% sub_cache_nodes/3
+%%
+%% Manages the substitutions record.
+
+%% Builds a new sub.
+-spec sub_new(term()) -> sub().
+sub_new(Temp) ->
+    #sub{t=Temp}.
+
+%% Folds the sub into a new one if the variables in nodes are not disjoint
+sub_fold(VarNames, #sub{v=SubNames}=Sub) ->
+    case is_disjoint_variables(VarNames, SubNames) of
+	true  -> Sub#sub{t={temp,Sub#sub.t}};
+	false -> sub_new({sub,Sub})
+    end.
+
+%% Unfolds the sub in case one was folded in the previous step
+sub_unfold(#sub{t={temp,Temp}}=Sub) ->
+    Sub#sub{t=Temp};
+sub_unfold(#sub{t={sub,Sub}}) ->
+    Sub.
+
+%% Adds the keys extracted from patterns to the state.
+-spec sub_add_keys([term()], sub()) -> {sub(),[term()]}.
+sub_add_keys(Keys0, #sub{p=Pat0}=Sub) ->
+    {Pat1,Keys1} = sub_add_keys(Keys0, Pat0, []),
+    {Sub#sub{p=Pat1},Keys1}.
+
+sub_add_keys([H|T], Pat, Keys) ->
+    case Pat of
+	#{H := _} ->
+	    %% We have already seen the pattern upstream
+	    sub_add_keys(T, Pat, Keys);
```

Comment:
Yes, not valid. I was unable to reach the line using the `error(reached)` trick (explained in another comment) when compiling all OTP applications.

I suggest that you add an assertion similar to this:

```

false = maps:is_key(H, Pat),        %Assertion.
sub_add_keys(T, maps:put(H, ?NOTSET, Pat), [H|Keys])

```